velox/substrait/SubstraitToVeloxExpr.cpp

```diff
@@ -0,0 +1,267 @@
+  }
+}
+
+std::shared_ptr<const core::ITypedExpr>
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression& sExpr,
+    const int32_t& inputPlanNodeId) {
+  std::shared_ptr<const core::ITypedExpr> veloxExpr;
+  auto typeCase = sExpr.rex_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression::RexTypeCase::kLiteral: {
+      veloxExpr = toVeloxExpr(sExpr.literal());
+      break;
+    }
+    case ::substrait::Expression::RexTypeCase::kScalarFunction: {
+      veloxExpr = toVeloxExpr(sExpr.scalar_function(), inputPlanNodeId);
+      break;
+    }
+    case ::substrait::Expression::RexTypeCase::kSelection: {
+      veloxExpr = toVeloxExpr(sExpr.selection(), inputPlanNodeId);
+      break;
+    }
+    default:
+      VELOX_NYI(
+          "Substrait conversion not supported for Expression '{}'", typeCase);
+  }
+  return veloxExpr;
+}
+
+// TODO: Support different types here.
+class SubstraitVeloxExprConverter::FilterInfo {
+ public:
+  FilterInfo() {}
+  // Used to set the left bound.
+  void setLeft(double left, bool isExclusive) {
+    left_ = left;
+    leftExclusive_ = isExclusive;
+    if (!isInitialized_) {
+      isInitialized_ = true;
+    }
+  }
+
+  // Used to set the right bound.
+  void setRight(double right, bool isExclusive) {
+    right_ = right;
+    rightExclusive_ = isExclusive;
+    if (!isInitialized_) {
+      isInitialized_ = true;
+    }
+  }
+
+  // Will fordis Null value if called once.
+  void forbidsNull() {
+    nullAllowed_ = false;
+    if (!isInitialized_) {
+      isInitialized_ = true;
+    }
+  }
+
+  // Return the initialization status.
+  bool isInitialized() {
+    return isInitialized_ ? true : false;
+  }
+
+  // The left bound.
+  std::optional<double> left_ = std::nullopt;
+  // The right bound.
+  std::optional<double> right_ = std::nullopt;
+  // The Null allowing.
+  bool nullAllowed_ = true;
+  // If true, left bound will be exclusive.
+  bool leftExclusive_ = false;
+  // If true, right bound will be exclusive.
+  bool rightExclusive_ = false;
+
+ private:
+  bool isInitialized_ = false;
+};
+
+void SubstraitVeloxExprConverter::getFlatConditions(
+    const ::substrait::Expression& sFilter,
+    std::vector<::substrait::Expression_ScalarFunction>* scalarFunctions) {
+  auto typeCase = sFilter.rex_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression::RexTypeCase::kScalarFunction: {
+      auto sFunc = sFilter.scalar_function();
+      auto filterName = subParser_->findSubstraitFunction(
+          functionMap_, sFunc.function_reference());
+      // TODO: Only AND relation is supported here.
+      if (filterName == "AND") {
+        for (const auto& sCondition : sFunc.args()) {
+          getFlatConditions(sCondition, scalarFunctions);
+        }
+      } else {
+        (*scalarFunctions).push_back(sFunc);
+      }
+      break;
+    }
+    default:
+      VELOX_NYI("GetFlatConditions not supported for type '{}'", typeCase);
```

Comment:
Is there a `toString()` sort of method for typeCase? I'm assuming this is a enum, so we would end up printing the integer value, which is not very helpful to users.

Reply:
Yes, the typeCase is just integer value. But after searching out, did not find an elegant way to map this integer value to the key word in enum. I found two options not quite simple:

- Construct a map to store the mapping relations
- Use a tool to map the integer back to key word: https://github.com/Neargye/magic_enum